### ISSUE TITLE
InvalidCastException when table merging

### ISSUE THREAD

**Issue by ichemidov** [linq2db/linq2db#1209]
I'm trying update table Registry with List of my own class objects. I found that I should use temp tables for merge with my collection and then merge this temp table with target table.
But when I use Merge operation I'm getting _InvalidCastException_.

```
Exception message: System.InvalidCastException: Specified cast is not valid.
Stack trace:
   at LinqToDB.Linq.Builder.AllJoinsLinqBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\AllJoinsLinqBuilder.cs:line 44
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 175
   at LinqToDB.Linq.Builder.ContextParser.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ContextParser.cs:line 22
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 175
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 146
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 280
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 233
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache) in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 84
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression) in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 165
   at LinqToDB.DataProvider.BasicMergeBuilder`2.BuildPredicateByKeys(Type keyType, Expression targetKey, Expression sourceKey) in C:\projects\linq2db\Source\LinqToDB\DataProvider\BasicMergeBuilder.cs:line 97
   at LinqToDB.DataProvider.BasicMergeBuilder`2.BuildMatch() in C:\projects\linq2db\Source\LinqToDB\DataProvider\BasicMergeBuilder.cs:line 561
   at LinqToDB.DataProvider.BasicMergeBuilder`2.BuildCommandText() in C:\projects\linq2db\Source\LinqToDB\DataProvider\BasicMergeBuilder.cs:line 646
   at LinqToDB.DataProvider.BasicMergeBuilder`2.BuildCommand() in C:\projects\linq2db\Source\LinqToDB\DataProvider\BasicMergeBuilder.cs:line 1265
   at LinqToDB.DataProvider.DataProviderBase.Merge[TTarget,TSource](DataConnection dataConnection, IMergeable`2 merge) in C:\projects\linq2db\Source\LinqToDB\DataProvider\DataProviderBase.cs:line 481
   at WsMVC.Core.ExpenditurePaper.Incur() in D:\Projects\ProjMVC\MVC\Core\Document.cs:line 225
```

### Steps to reproduce

```c#
public class DocumentContent
    {
        public int CardId { get; set; }
        public double Quantity { get; set; }
    }
```

```c#
[Table(Name = "Registry")]
    public class Registry
    {
        [Column(Name = "CardId"), PrimaryKey, Identity]
        public int CardId{ get; set; }
        [Column(Name = "Quantity"), NotNull]
        public double Quantity { get; set; }
}
```

```c#
var tsource = Content.GroupBy(l => l.CardId).Select(s => new DocumentContent
    {
       CardId = s.Key,
       Quantity = s.Sum(c=> c.Quantity)
    }).ToList();
var tempTable = db.CreateTable<DocumentContent>("#tmp");
tempTable.BulkCopy(tsource);
var source = tempTable.Select(s => s);

db.GetTable<Registry>()
.Merge()
.Using(source)
.On(s => s.CardId, f => f.CardId)
.UpdateWhenMatched((t, s) => new Registry()
{
   Quantity = t.Quantity + s.Quantity
})
.Merge();
```

### Environment details
linq2db version: *2.1.0*
Database Server: *SqlServer 2016 (13.0.4001.0)*
Operating system: *Windows 8*
Framework version: *.NET Framework 4.5.2*

**Comment by sdanyliv** [comment]
@ichemidov, will handle this. Thanks for reporting.

**Comment by sdanyliv** [comment]
BTW, it is syntetic sample? I do not see purpose why you have used temporary table. Grouping query can be used in merge operation without objects materialzation.

**Comment by ichemidov** [comment]
In some ways it's syntetic sample. I've create it to understand how works merge api. But In the future I plan to extend it (add fields to classes in sample) and use in my project. 
I found that I should use temp tables in issue [#277](https://github.com/linq2db/linq2db/issues/277#issuecomment-151540140), beacuse I want merge database table with `List<DocumentContent>` (no such table exists in db, it's only class in my programm).

**Comment by sdanyliv** [comment]
@ichemidov, unfortunately i can not reproduce issue. 
Do you have some extra code that transforms Expression Tree?

**Comment by sdanyliv** [comment]
I ask because such exception is possible if somewhere in dynamic Expression Tree generation used incorrect Join method call. I have reviewed our code according to call stack and not found problems.
If you can supply us sample project it will be great. 
Thanks!

**Comment by ichemidov** [comment]
I created [sample project](https://github.com/linq2db/linq2db/files/2154217/WsCore.zip) and I was even able to localize the problem.
See `WsDataProvider.cs`

No Exception
```
var cont = DocsCont.Where(s => s.IdDoc == id).Select(s => new { s.CardId, s.Quantity });
foreach (var item in cont)
 {
    var dc = new DocumentContent
    {
        CardId = (int)item.CardId,
        ResourceId = 0,
        Code = "",
        Quantity = (double)item.Quantity
    };
    contentList.Add(dc);
}
```

InvalidCastException 
```
var cont = DocsCont.Where(s => s.IdDoc == id).InnerJoin(Resource,
(s, e) => s.ResourceId == e.ResourceId,
(s, e) => new { s.CardId, e.ResourceId, e.Code, s.Quantity });
foreach (var item in cont)
{
    var dc = new DocumentContent
    {
        CardId = item.CardId,
        ResourceId = item.ResourceId,
        Code = item.Code,
        Quantity = (double)item.Quantity
    };
    contentList.Add(dc);
}
```

**Comment by sdanyliv** [comment]
@ichemidov, again. 
It works for me.  Looks like you have some post build plugin that modifies Expression Tree. 
If yes. Let me know which one. 
We have introduced our `Join` method, it has a little bit different parameters that IQueryable equivalent and looks like some extension won't work if expression tree contains this method.
```cs
public static IQueryable<TResult> Join<TOuter, TInner, TResult>(
	[NotNull]           this IQueryable<TOuter>                   outer,
	[NotNull]           IQueryable<TInner>                        inner,
	[SqlQueryDependent] SqlJoinType                               joinType,
	[NotNull, InstantHandle] Expression<Func<TOuter, TInner, bool>>    predicate,
	[NotNull, InstantHandle] Expression<Func<TOuter, TInner, TResult>> resultSelector)
```
`InnerJoin` call in your sample translated to this `Join` function with parameter `SqlJoinType.Inner`

P.S.
Simple comment for your sample solution. Better to generate models by [T4 templates](https://github.com/linq2db/linq2db/tree/master/Source/LinqToDB.Templates#t4-models)
And create methods like `GetWsDocumentById` using extension methods.

**Comment by sdanyliv** [comment]
Really strange. Could you please clone linq2db sources and add refrence to the library in your test project?

**Comment by ichemidov** [comment]
So, I cloned linq2db sources and build it (I had to change file `\Source\LinqToDB\Compile.cmd` to specify the path to MSBuild.exe), and add refere to the .dll in test project.
I got InvalidCastException...
Github does not allow to attach file (too big)
https://yadi.sk/d/TTI5IObT3YhfZk

**Comment by sdanyliv** [comment]
Just add reference to the linq2db.csproj. And try figure our why it fails.
Probjem with Join, but why...

**Comment by sdanyliv** [comment]
Fixed in bb1ef44f231151195c2c0be8bd844606c0e8d542
